### PR TITLE
raft: don't ack MsgPreVotes when it cannot succeed

### PR DESCRIPTION
**Pull request by xiang90**
replace https://github.com/coreos/etcd/pull/8517.

i added a test, which illustrate what problem this pr tries to address.  

/cc @bdarnell @javaforfun

### CODE REVIEW

**Comment by javaforfun** [comment]
the test success even without the change in raft.go

**Comment by xiang90** [comment]
@javaforfun The test case is added for ensuring the correctness (I can add a test to make the previous case not passing though). The current code is added for clarity.

The next improvement should be changing the condition to something like

```
((m.Type == pb.MsgVote && (r.Vote == None || r.Vote == m.From)) ||
			(m.Type == pb.MsgPreVote && m.Term > r.Term)) &&
			r.raftLog.isUpToDate(m.Index, m.LogTerm) 
```

This change requires more test changes.

**Comment by xiang90** [comment]
i am trying to merge in the change #8517 made first.

File: raft/raft_test.go [hunk 1]
```diff
@@ -3273,6 +3273,60 @@ func TestPreVoteWithSplitVote(t *testing.T) {
...
+
+	// check whether the term values are expected
+	// n3.Term == 3
+	sm = nt.peers[3].(*raft)
+	if sm.Term != 3 {
+		t.Errorf("peer 3 term: %d, want %d", sm.Term, 3)
+	}
+
+	// check state
+	// n3 == follower
```
Comment (bdarnell) [inline raft/raft_test.go]: This doesn't tell us whether the pre-vote was granted (I think). The pre-vote could have been granted and then the final vote rejected. I'm not sure how to test this properly unless we record all messages sent and verify that node 3 never sends any MsgVotes.

**Pull request closed without merging**